File: src/main/java/com/lessmarkup/userinterface/nodehandlers/common/TabPageNodeHandler.java
package com.lessmarkup.userinterface.nodehandlers.common;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import com.lessmarkup.framework.helpers.JsonSerializer;
import com.lessmarkup.framework.helpers.StringHelper;
import com.lessmarkup.framework.nodehandlers.AbstractNodeHandler;
import com.lessmarkup.interfaces.cache.DataCache;
import com.lessmarkup.interfaces.security.CurrentUser;
import com.lessmarkup.interfaces.structure.*;
import com.lessmarkup.userinterface.model.structure.LoadNodeViewModel;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.OptionalLong;

public abstract class TabPageNodeHandler extends AbstractNodeHandler {
    static class TabPage {
        private String title;
        private Class<? extends NodeHandler> type;
        private JsonObject viewData;
        private String viewBody;
        private String path;
        private String fullPath;
        private OptionalLong pageId;
        private String settings;
        private Optional<NodeAccessType> accessType;
        private String uniqueId;

        public String getTitle() {
            return title;
        }

        public void setTitle(String title) {
            this.title = title;
        }

        public Class<? extends NodeHandler> getType() {
            return type;
        }

        public void setType(Class<? extends NodeHandler> type) {
            this.type = type;
        }

        public JsonObject getViewData() {
            return viewData;
        }

        public void setViewData(JsonObject viewData) {
            this.viewData = viewData;
        }

        public String getViewBody() {
            return viewBody;
        }

        public void setViewBody(String viewBody) {
            this.viewBody = viewBody;
        }

        public String getPath() {
            return path;
        }

        public void setPath(String path) {
            this.path = path;
        }

        public String getFullPath() {
            return fullPath;
        }

        public void setFullPath(String fullPath) {
            this.fullPath = fullPath;
        }

        public OptionalLong getPageId() {
            return pageId;
        }

        public void setPageId(OptionalLong pageId) {
            this.pageId = pageId;
        }

        public String getSettings() {
            return settings;
        }

        public void setSettings(String settings) {
            this.settings = settings;
        }

        public Optional<NodeAccessType> getAccessType() {
            return accessType;
        }

        public void setAccessType(Optional<NodeAccessType> accessType) {
            this.accessType = accessType;
        }

        public String getUniqueId() {
            return uniqueId;
        }

        public void setUniqueId(String uniqueId) {
            this.uniqueId = uniqueId;
        }
    }

    private final List<TabPage> pages = new ArrayList<>();
    private final DataCache dataCache;
    private final CurrentUser currentUser;
    private final List<String> scripts = new ArrayList<>();

    protected TabPageNodeHandler(DataCache dataCache, CurrentUser currentUser) {
        this.currentUser = currentUser;
        this.dataCache = dataCache;
    }

    protected <T extends NodeHandler> void addPage(Class<T> type, String title, String path) {
        TabPage page = new TabPage();
        page.setType(type);
        page.setPath(path);
        page.setTitle(title);
        pages.add(page);
    }

    protected <T extends NodeHandler> void addPage(Class<T> type, String title, String path, OptionalLong pageId) {
        TabPage page = new TabPage();
        page.setType(type);
        page.setPath(path);
        page.setTitle(title);
        page.setPageId(pageId);
        pages.add(page);
    }

    @Override
    protected Object initialize() {
        NodeCache nodeCache = dataCache.get(NodeCache.class);

        String parentPath = "";

        if (getObjectId().isPresent()) {
            CachedNodeInformation currentNode = nodeCache.getNode(getObjectId().getAsLong());
            parentPath = currentNode.getFullPath();

            for (CachedNodeInformation child : currentNode.getChildren()) {
                NodeAccessType accessType = child.checkRights(currentUser);

                if (accessType == NodeAccessType.NO_ACCESS) {
                    continue;
                }

                TabPage page = new TabPage();
                page.setType(child.getHandlerType());
                page.setPageId(OptionalLong.of(child.getNodeId()));
                page.setSettings(child.getSettings());
                page.setPath(child.getPath());
                page.setFullPath(child.getFullPath());
                page.setTitle(child.getTitle());
                page.setAccessType(Optional.of(accessType));

                pages.add(page);
            }
        }

        pages.sort((p1, p2) -> p1.getTitle().compareTo(p2.getTitle()));

        int unknownPath = 1;
        int uniqueId = 1;

        for (TabPage page : pages) {
            NodeHandler handler = createChildHandler(page.getType());

            JsonElement nodeSettings = null;

            if (handler.getSettingsModel() != null && !StringHelper.isNullOrEmpty(page.getSettings())) {
                nodeSettings = JsonSerializer.deserializeToTree(page.getSettings());
            }

            if (StringHelper.isNullOrWhitespace(page.getPath())) {
                page.setPath(String.format("page_%s", unknownPath++));
                page.setFullPath(page.getPath());

                if (!StringHelper.isNullOrEmpty(parentPath)) {
                    page.setFullPath(parentPath + "/" + page.getPath());
                }
            }

            handler.initialize(page.getPageId(), 
                    nodeSettings != null && nodeSettings.isJsonObject() ? nodeSettings.getAsJsonObject() : null, 
                    page.getPath(), page.getFullPath(), page.getAccessType().isPresent() ? page.getAccessType().get() : getAccessType());

            page.setViewData(handler.getViewData());
            page.setViewBody(LoadNodeViewModel.getViewTemplate(handler, dataCache));
            page.setUniqueId(String.format("page_%s", uniqueId++));

            List<String> handlerScripts = handler.getScripts();
            if (handlerScripts != null) {
                this.scripts.addAll(handlerScripts);
            }
        }

        pages.removeIf(p -> p.getViewBody() == null);

        return null;
    }

    @Override
    public JsonObject getViewData() {
        JsonObject ret = new JsonObject();
        JsonArray pagesArray = new JsonArray();
        this.pages.forEach(p -> {
            JsonObject o = new JsonObject();
            o.addProperty("pageId", p.getPageId().orElse(0));
            o.addProperty("path", p.getPath());
            o.addProperty("title", p.getTitle());
            o.addProperty("viewBody", p.getViewBody());
            o.add("viewData", p.getViewData());
            o.addProperty("uniqueId", p.getUniqueId());
            pagesArray.add(o);
        });
        ret.add("pages", pagesArray);
        JsonArray requires = new JsonArray();
        this.scripts.forEach(s -> requires.add(new JsonPrimitive(s)));
        ret.add("requires", requires);
        return ret;
    }

    @Override
    public String getViewType() {
        return "TabPage";
    }

    @Override
    public boolean hasChildren() {
        return true;
    }

    @Override
    public ChildHandlerSettings getChildHandler(String path) {
        Optional<TabPage> page = pages.stream().filter(p -> p.getPath().equals(path)).findFirst();
        if (!page.isPresent()) {
            return null;
        }
        NodeHandler handler = createChildHandler(page.get().getType());

        JsonElement nodeSettings = null;
        if (handler.getSettingsModel() != null && !StringHelper.isNullOrEmpty(page.get().getSettings())) {
            nodeSettings = JsonSerializer.deserializeToTree(page.get().getSettings());
        }

        handler.initialize(page.get().getPageId(), 
                nodeSettings != null && nodeSettings.isJsonObject() ? nodeSettings.getAsJsonObject() : null, 
                page.get().getPath(), page.get().getFullPath(), page.get().getAccessType().isPresent() ? page.get().getAccessType().get() : getAccessType());

        ChildHandlerSettings ret = new ChildHandlerSettings();
        ret.setHandler(handler);
        ret.setPath(path);
        ret.setTitle(page.get().getTitle());
        ret.setId(page.get().getPageId());

        return ret;
    }
}

File: src/main/scala/com/lessmarkup/userinterface/nodehandlers/common/TabPageNodeHandler.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.userinterface.nodehandlers.common

import com.google.gson.{JsonArray, JsonObject, JsonPrimitive}
import com.lessmarkup.framework.helpers.{JsonSerializer, StringHelper}
import com.lessmarkup.framework.nodehandlers.{AbstractNodeHandler, NodeHandlerConfiguration}
import com.lessmarkup.framework.system.RequestContextHolder
import com.lessmarkup.interfaces.annotations.NodeAccessType
import com.lessmarkup.interfaces.cache.DataCache
import com.lessmarkup.interfaces.structure._
import com.lessmarkup.userinterface.model.structure.LoadNodeViewModel
import com.lessmarkup.userinterface.nodehandlers.common.TabPageNodeHandler.TabPageWithData

object TabPageNodeHandler {
  class TabPage(
      val title: String,
      val handlerType: Class[_ <: NodeHandlerFactory],
      val viewData: Option[JsonObject] = None,
      val viewBody: Option[String] = None,
      val path: String,
      val fullPath: String,
      val uniqueId: Option[String] = None,
      val pageId: Option[Long] = None,
      val settings: String = "",
      val accessType: Option[NodeAccessType] = None)

  protected class TabPageWithData(val configuration: TabPage, val handler: NodeHandler, val viewData: Option[JsonObject], val viewBody: String)
}

abstract class TabPageNodeHandler(dataCache: DataCache, configuration: NodeHandlerConfiguration) extends AbstractNodeHandler(configuration) {

  private val pages = createPages
  private val createdPages = createPageHandlers
  private val scripts = createdPages.flatMap(_.handler.getScripts)

  protected def createPages: Seq[TabPageNodeHandler.TabPage] = {
    val nodeCache: NodeCache = dataCache.get(classOf[NodeCache])

    if (getObjectId.isEmpty) {
      return Nil
    }

    val currentNode: CachedNodeInformation = nodeCache.getNode(getObjectId.get).get

    //val parentPath = currentNode.getFullPath

    val pages = for (
      child <- currentNode.children;
      accessType: NodeAccessType = child.checkRights(RequestContextHolder.getContext.getCurrentUser)
      if accessType != NodeAccessType.NO_ACCESS && child.handlerType.isDefined
    ) yield {
      new TabPageNodeHandler.TabPage(
        handlerType = child.handlerType.get,
        pageId = Option(child.nodeId),
        settings = child.settings,
        path = child.path,
        fullPath = child.fullPath,
        title = child.title,
        accessType = Option(accessType)
      )
    }

    pages.sortBy(_.title)
  }

  private def createPageHandlers: Seq[TabPageNodeHandler.TabPageWithData] = {
    val createdPages: Seq[Option[TabPageWithData]] = for ((page, index) <- pages zipWithIndex) yield {

      val nodeSettings = if (!StringHelper.isNullOrEmpty(page.settings)) {
          Option(JsonSerializer.deserializeToTree(page.settings).getAsJsonObject)
        } else {
          None
        }

      val path = if (StringHelper.isNullOrWhitespace(page.path)) s"page_$index" else page.path

      val nodeHandlerConfiguration = new NodeHandlerConfiguration(
        objectId = Option(index.toLong),
        settings = nodeSettings,
        accessType = page.accessType.getOrElse(getAccessType),
        path = path,
        fullPath = path
      )

      val handler: NodeHandler = createChildHandler(page.handlerType, nodeHandlerConfiguration)

      val viewBody = LoadNodeViewModel.getViewTemplate(handler, dataCache)

      if (viewBody == null) {
        None
      } else {

        val viewData = handler.getViewData

        val pageWithData = new TabPageWithData(
          configuration = page,
          handler = handler,
          viewData = viewData,
          viewBody = viewBody
        )

        Option(pageWithData)
      }
    }

    createdPages.flatten
  }

  protected def createPage[T <: NodeHandlerFactory](handlerType: Class[T], title: String, path: String): TabPageNodeHandler.TabPage = {
    new TabPageNodeHandler.TabPage(
      handlerType = handlerType,
      path = path,
      fullPath = path,
      title = title
    )
  }

  override def getViewData: Option[JsonObject] = {
    val ret: JsonObject = new JsonObject
    val pagesArray: JsonArray = new JsonArray
    createdPages.foreach(p => {
      val o = new JsonObject()
      val c = p.configuration
      o.addProperty("pageId", c.pageId.getOrElse(0L))
      o.addProperty("path", c.path)
      o.addProperty("title", c.title)
      o.addProperty("viewBody", p.viewBody)
      o.add("viewData", p.viewData.orNull)
      o.addProperty("uniqueId", c.uniqueId.get)
      pagesArray.add(o)
    })
    ret.add("pages", pagesArray)
    val requires: JsonArray = new JsonArray
    scripts.foreach(s => requires.add(new JsonPrimitive(s)))
    ret.add("requires", requires)
    Option(ret)
  }

  override def getViewType: String = {
    "TabPage"
  }

  override def hasChildren: Boolean = {
    true
  }

  override def createChildHandler(path: Seq[String]): Option[ChildHandlerSettings] = {
    val pathCombined = path.mkString("/")
    val page = createdPages.find(_.configuration.path == pathCombined)
    if (page.isEmpty) {
      return None
    }
    val nodeSettings = if (!StringHelper.isNullOrEmpty(page.get.configuration.settings)) {
      Option(JsonSerializer.deserializeToTree(page.get.configuration.settings).getAsJsonObject)
    } else {
      None
    }

    val nodeHandlerConfiguration = new NodeHandlerConfiguration(
      objectId = page.get.configuration.pageId,
      settings = nodeSettings,
      accessType = page.get.configuration.accessType.getOrElse(getAccessType),
      path = page.get.configuration.path,
      fullPath = page.get.configuration.fullPath
    )

    val handler: NodeHandler = createChildHandler(page.get.configuration.handlerType, nodeHandlerConfiguration)

    Option(new ChildHandlerSettings(
      handler = handler,
      id = page.get.configuration.pageId,
      title = page.get.configuration.title,
      path = page.get.configuration.path,
      rest = Seq.empty
    ))
  }
}
